File: src/main/java/toby/command/commands/music/JoinCommand.java
package toby.command.commands.music;

import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.GuildVoiceState;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.channel.middleman.AudioChannel;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.managers.AudioManager;
import toby.command.CommandContext;
import toby.jpa.dto.ConfigDto;
import toby.jpa.dto.UserDto;
import toby.jpa.service.IConfigService;
import toby.lavaplayer.PlayerManager;

import static toby.command.ICommand.invokeDeleteOnMessageResponse;

public class JoinCommand implements IMusicCommand {
    private final IConfigService configService;

    public JoinCommand(IConfigService configService) {
        this.configService = configService;
    }

    @Override
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        handleMusicCommand(ctx, PlayerManager.getInstance(), requestingUserDto, deleteDelay);
    }

    @Override
    public void handleMusicCommand(CommandContext ctx, PlayerManager instance, UserDto requestingUserDto, Integer deleteDelay) {
        final SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        final Member self = ctx.getSelfMember();

        if (!requestingUserDto.hasMusicPermission()) {
            sendErrorMessage(event, deleteDelay);
            return;
        }

        final GuildVoiceState memberVoiceState = doJoinChannelValidation(ctx, deleteDelay);
        if (memberVoiceState == null) return;

        final AudioManager audioManager = event.getGuild().getAudioManager();
        final AudioChannel memberChannel = memberVoiceState.getChannel();

        if (self.hasPermission(Permission.VOICE_CONNECT)) {
            audioManager.openAudioConnection(memberChannel);
            String volumePropertyName = ConfigDto.Configurations.VOLUME.getConfigValue();
            ConfigDto databaseConfig = configService.getConfigByName(volumePropertyName, event.getGuild().getId());
            int defaultVolume = databaseConfig != null ? Integer.parseInt(databaseConfig.getValue()) : 100;
            instance.getMusicManager(event.getGuild()).getAudioPlayer().setVolume(defaultVolume);
            event.getHook().sendMessageFormat("Connecting to `\uD83D\uDD0A %s` with volume '%s'", memberChannel.getName(), defaultVolume).queue(invokeDeleteOnMessageResponse(deleteDelay));
        }
    }

    private GuildVoiceState doJoinChannelValidation(CommandContext ctx, Integer deleteDelay) {
        final Member self = ctx.getSelfMember();
        final GuildVoiceState selfVoiceState = self.getVoiceState();
        SlashCommandInteractionEvent event = ctx.getEvent();

        if (selfVoiceState.inAudioChannel()) {
            event.getHook().sendMessage("I'm already in a voice channel").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay));
            return null;
        }

        final Member member = ctx.getMember();
        final GuildVoiceState memberVoiceState = member.getVoiceState();

        if (!memberVoiceState.inAudioChannel()) {
            event.getHook().sendMessage("You need to be in a voice channel for this command to work").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay));
            return null;
        }
        return memberVoiceState;
    }

    @Override
    public String getName() {
        return "join";
    }

    @Override
    public String getDescription() {
        return "Makes the bot join your voice channel";
    }
}

File: src/main/kotlin/toby/command/commands/music/JoinCommand.kt
package toby.command.commands.music

import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.GuildVoiceState
import toby.command.CommandContext
import toby.command.ICommand.Companion.invokeDeleteOnMessageResponse
import toby.handler.Handler
import toby.jpa.dto.ConfigDto
import toby.jpa.dto.UserDto
import toby.jpa.service.IConfigService
import toby.lavaplayer.PlayerManager

class JoinCommand(private val configService: IConfigService) : IMusicCommand {

    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int?) {
        handleMusicCommand(ctx, PlayerManager.instance, requestingUserDto, deleteDelay)
    }

    override fun handleMusicCommand(ctx: CommandContext, instance: PlayerManager, requestingUserDto: UserDto, deleteDelay: Int?) {
        val event = ctx.event
        event.deferReply().queue()

        val self = ctx.selfMember ?: return
        val memberVoiceState = doJoinChannelValidation(ctx, deleteDelay) ?: return
        val audioManager = event.guild!!.audioManager
        val memberChannel = memberVoiceState.channel?.asVoiceChannel()

        if (!requestingUserDto.musicPermission) {
            sendErrorMessage(event, deleteDelay!!)
            return
        }

        if (!self.hasPermission(Permission.VOICE_CONNECT)) {
            sendErrorMessage(event, deleteDelay!!)
            return
        }

        audioManager.openAudioConnection(memberChannel)
        Handler.lastConnectedChannel[event.guild!!.idLong] = memberChannel!!
        val volumePropertyName = ConfigDto.Configurations.VOLUME.configValue
        val databaseConfig = configService.getConfigByName(volumePropertyName, event.guild?.id)
        val defaultVolume = databaseConfig?.value?.toInt() ?: 100
        instance.getMusicManager(event.guild!!).audioPlayer.volume = defaultVolume

        event.hook.sendMessage("Connecting to `\uD83D\uDD0A ${memberChannel.name}` with volume '$defaultVolume'").queue(invokeDeleteOnMessageResponse(deleteDelay!!))
    }

    private fun doJoinChannelValidation(ctx: CommandContext, deleteDelay: Int?): GuildVoiceState? {
        val self = ctx.selfMember ?: return null
        val selfVoiceState = self.voiceState
        val event = ctx.event

        if (selfVoiceState?.inAudioChannel() == true) {
            event.hook.sendMessage("I'm already in a voice channel").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay!!))
            return null
        }

        val member = ctx.member ?: return null
        val memberVoiceState = member.voiceState

        if (memberVoiceState?.inAudioChannel() != true) {
            event.hook.sendMessage("You need to be in a voice channel for this command to work").setEphemeral(true).queue(invokeDeleteOnMessageResponse(deleteDelay!!))
            return null
        }

        return memberVoiceState
    }

    override val name: String = "join"
    override val description: String = "Makes the bot join your voice channel"
}
